engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -214,43 +209,43 @@ boolean equalsStrings(Object selfString, Object otherString,
     } catch (UnsupportedMessageException e) {
       throw new IllegalStateException(e);
     }
-    return Normalizer.compare(
-        selfJavaString,
-        otherJavaString,
-        Normalizer.FOLD_CASE_DEFAULT
-    ) == 0;
+    return Normalizer.compare(selfJavaString, otherJavaString, Normalizer.FOLD_CASE_DEFAULT) == 0;
   }
 
-  @Specialization(
-    guards = "isPrimitive(self, strings) != isPrimitive(other, strings)"
-  )
-  boolean equalsDifferent(Object self, Object other, @CachedLibrary(limit = "10") InteropLibrary strings) {
+  @Specialization(guards = "isPrimitive(self, interop) != isPrimitive(other, interop)")
+  boolean equalsDifferent(
+      Object self, Object other, @CachedLibrary(limit = "10") InteropLibrary interop) {
     return false;
   }
 
   /** Equals for Atoms and AtomConstructors */
-
   @Specialization
   boolean equalsAtomConstructors(AtomConstructor self, AtomConstructor other) {
     return self == other;
   }
 
   @Specialization
-  boolean equalsAtoms(Atom self, Atom other, @Cached EqualsAtomNode equalsNode) {
-    return equalsNode.execute(self, other);
+  boolean equalsAtoms(
+      Atom self,
+      Atom other,
+      @Cached EqualsAtomNode equalsNode,
+      @Cached IsSameObjectNode isSameObjectNode) {
+    return isSameObjectNode.execute(self, other) || equalsNode.execute(self, other);
```

Comment:
This is the fix for #6065, right?

Reply:
Yes, #6065 is addressed in these commits:
- 35074849afa320cac15d866b9c7f3921ef3f2bc4
- 55ceaba9b1ff2a0cf63fec5b11aabbad834c0e06